Move Linux frame buffer support to base system

wulf7 opened the pull request:
On Linux, framebuffer code is not a part of drm and can be used with non-drm drivers.
The aim is to move linux_fb.c vt_drmfb.c to linuxkpi_video module and import to base system after than.
The changes clashes with ongoing drm-kmod 6.xx import doing by @dumbbell. That should be resolved in future.
Still WIP

wulf7 committed:
vt_drmfb: drop dependency on struct drm_fb_helper

Generic framebufer driver is a lower layer than drm framebufer.
It should not depend on later.
The only usage of passing `struct drm_fb_helper` to vt_drmfb was using
it as a container of reference to parent BSD device.
Pass device directly from drm to vt_drmfb to not depend on drm.

Sponsored by:	Future Crew LLC

wulf7 committed:
Move skip_ddb global from drm to vt_drmfb

as vt_drmfb is the only user of skip_ddb. Keep skip_ddb sysctl
declarations in drm as they are children of drm sysctl nodes.

Sponsored by:	Future Crew LLC

wulf7 committed:
Move registration of fictitious page ranges back to framebufer

support code to consolidate it in one place. This done with adding of
FreeBSD-specific fields to drm_driver structure to pass aperture
parametrs from drm drivers to framebufer support code.

Also remove struct apertures_struct from fb_info to follow upstream.

Sponsored by:	Future Crew LLC

wulf7 committed:
drivers/gpu/drm/drm_os_freebsd.[ch]: Minor cleanup

Remove references to now unused headers, functions and kernel modules.

Sponsored by:	Future Crew LLC

wulf7 committed:
linux/fb.h: Simplify deferred I/O - related macros

and move them to BSD-licensed linux/fb.h

Sponsored by:	Future Crew LLC

wulf7 committed:
Move sys_ functions from linux_fb.c to BSD linux/fb.h

While here move cfb_ declarations to BSD linux/fb.h too and
remove extraneous extern keywords.

Sponsored by:	Future Crew LLC

emaste commented:
> The aim is to move linux_fb.c vt_drmfb.c to linuxkpi_video module and import to base system after than.

Would be good to also indicate why they should be in the base system

wulf7 commented:
We a going to make drm.ko to be a part of base system

wulf7 committed:
Move remnants of GPL-licensed linux/fb.h to BSD-licensed one

wulf7 committed:
Delete GPL-licensed linux/fb.h

wulf7 committed:
linux_fb.c: Add linuxkpi prefix to all exportable functions

to match LinuxKPI style.

wulf7 committed:
Move linux_fb.c and vt_drmfb.[ch] from drm.ko to linux_video.ko

They do not depend on drm internals.
Add -DDEVELOPER to make options to compile linux_video.ko

wulf7 committed:
linux_fb.c: Fix build after moving linux_fb.c to linux_video

wulf7 committed:
linuxkpi_video: Use aperture_remove_conflicting_(pci_)devices from linux_fb.c

They are native.

wulf7 committed:
vt_drmfb.c: Make all vt_driver members static

wulf7 committed:
vt_drmfb: Do not use wrappers to call vt_suspend and vt_resume

That is like most other VT drivers do.

wulf7 committed:
Move skip_ddb declaration from vt_drmfb.h to linux/fb.h

Rename it to linuxkpi_skip_ddb as it is public symbol now.

wulf7 committed:
Do not panic if aperture_base or aperture_size is not defined

wulf7 committed:
linux/fb.h: Remove extraneous porting from struct fb_fix_screeninfo

This structure is used only in LKPI so no reasons to convert members
to native types.

wulf7 committed:
vt_drmfb: Convert to kernel module compatible with system fbd

system fbd can be enabled with setting of loader tunable
kern.vt.disable_drmfb to 1

wulf7 committed:
vt_drmfb: Convert to generic postswitch handler

To share it with system fbc

wulf7 committed:
linux_fb.c: Remove PCI_STD_NUM_BARS definition

It is already in linux/pci.h

wulf7 committed:
linux/fb.h: Add dummy fb_deferred_io_cleanup() implementation

wulf7 committed:
linux_fb: do not truncate bitmap in vt_drmfb bitblt

That simplifies the code.

wulf7 committed:
vt_drmfb: Inherit vt_fb init and fini methods

instead rolling of own copies. They are almost identical. The only
difference is extra postswitch call in init method of vt_fb.

wulf7 committed:
linux_fb: allocate fbd unit number automaticaly with DEVICE_UNIT_ANY

to avoid clashes with udl(4) driver.

wulf7 committed:
vt_drmfb: sort #includes

wulf7 committed:
linuxkpi_video: rename files to match LinuxKPI

wulf7 committed:
linux_hdmi.c: Drop remnants of drm dependency

wulf7 commented:
The MR is submitted to review: https://reviews.freebsd.org/D57488
